File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/7. 2667.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const N = +inputArr[0]
const board = inputArr.slice(1).map(row => row.trim().split('').map(Number))

// 방문 여부 저장
const isVisit = Array.from({length: N}, () => Array.from({length: N}).fill(false))

// 단지에 속한 집의 수를 저장하는 배열
const resultArr = []

// 모든 점마다 dfs 를 수행할 수 있는지 확인
for (let row = 0; row < N; row++) {
  for (let col = 0; col < N; col++) {
    if (board[row][col] === 1 && isVisit[row][col] === false) {
      const result = dfs(row, col)
      resultArr.push(result)
    }
  }
}

resultArr.sort((a, b) => a - b)

let resultStr = `${resultArr.length}\n`
resultArr.forEach(result => (resultStr += result + '\n'))

process.stdout.write(resultStr)

function dfs(row, col) {
  isVisit[row][col] = true
  let result = 1

  if (row > 0 && board[row - 1][col] === 1 && isVisit[row - 1][col] === false) {
    result += dfs(row - 1, col)
  }
  if (row < N - 1 && board[row + 1][col] === 1 && isVisit[row + 1][col] === false) {
    result += dfs(row + 1, col)
  }
  if (col > 0 && board[row][col - 1] === 1 && isVisit[row][col - 1] === false) {
    result += dfs(row, col - 1)
  }
  if (col < N - 1 && board[row][col + 1] === 1 && isVisit[row][col + 1] === false) {
    result += dfs(row, col + 1)
  }

  return result
}

File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/7. 2667.py
from collections import deque
import sys

input = sys.stdin.readline
output = sys.stdout.write

N = int(input())

board = list(list(map(int, list(input().strip()))) for i in range(N))

isVisit = list(list(False for i in range(N)) for j in range(N))

resultArr = []
visitQueue = deque()

def bfs(row, col):
  global N
  result = 1
  if row > 0 and board[row - 1][col] == 1 and not isVisit[row - 1][col]:
    isVisit[row - 1][col] = True
    visitQueue.append([row - 1, col])
  if row < N - 1 and board[row + 1][col] == 1 and not isVisit[row + 1][col]:
    isVisit[row + 1][col] = True
    visitQueue.append([row + 1, col]) 
  if col > 0 and board[row][col - 1] == 1 and not isVisit[row][col - 1]:
    isVisit[row][col - 1] = True
    visitQueue.append([row, col - 1])
  if col < N - 1 and board[row][col + 1] == 1 and not isVisit[row][col + 1]:
    isVisit[row][col + 1] = True
    visitQueue.append([row, col + 1])

  if len(visitQueue) > 0:
    nextRow, nextCol = visitQueue.popleft()
    result += bfs(nextRow, nextCol)

  return result

for row in range(N):
  for col in range(N):
    if board[row][col] == 1 and not isVisit[row][col]:
      isVisit[row][col] = True
      result = bfs(row, col)
      resultArr.append(result)

resultArr.sort()

output(f"{len(resultArr)}\n")
for result in resultArr:
  output(f"{result}\n")